diff --git a/cmd/bootm.c b/cmd/bootm.c
index 796d5e23adb..a8ba34cb5a1 100644
--- a/cmd/bootm.c
+++ b/cmd/bootm.c
@@ -207,6 +207,15 @@ int do_bootm(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 	case IMAGE_FORMAT_ANDROID:
 		/* Do this authentication in boota command */
 		break;
+#endif
+#ifdef CONFIG_FIT
+	case IMAGE_FORMAT_FIT:
+		if (authenticate_image(image_load_addr,
+			fit_get_size((void*)image_load_addr)) != 0) {
+			printf("Authenticate FIT image Fail, Please check\n");
+			return 1;
+		}
+		break;
 #endif
 	default:
 		printf("Not valid image format for Authentication, Please check\n");
