diff --git a/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c b/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
index a4fddda783f8..6cf21c5e05d6 100644
--- a/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
+++ b/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
@@ -63,9 +63,11 @@ static int ds070boe30t_panel_prepare(struct drm_panel *panel)
 	}
 
 	msleep(20);
-	gpiod_set_value(ctx->gpios.power, 1);
+	if (!IS_ERR(ctx->gpios.power))
+		gpiod_set_value(ctx->gpios.power, 1);
 	msleep(20);
-	gpiod_set_value(ctx->gpios.reset, 1);
+	if (!IS_ERR(ctx->gpios.reset))
+		gpiod_set_value(ctx->gpios.reset, 1);
 	msleep(20);
 
 	ctx->prepared = true;
@@ -80,8 +82,10 @@ static int ds070boe30t_panel_unprepare(struct drm_panel *panel)
 	if (!ctx->prepared)
 		return 0;
 
-	gpiod_set_value(ctx->gpios.reset, 0);
-	gpiod_set_value(ctx->gpios.power, 0);
+	if (!IS_ERR(ctx->gpios.reset))
+		gpiod_set_value(ctx->gpios.reset, 0);
+	if (!IS_ERR(ctx->gpios.power))
+		gpiod_set_value(ctx->gpios.power, 0);
 	regulator_disable(ctx->supply);
 
 	ctx->prepared = false;
@@ -323,14 +327,12 @@ static int ds070boe30t_panel_dsi_probe(struct mipi_dsi_device *dsi)
 
 	ctx->gpios.reset = devm_gpiod_get(&dsi->dev, "reset", GPIOD_OUT_LOW);
 	if (IS_ERR(ctx->gpios.reset)) {
-		dev_err(&dsi->dev, "Couldn't get our reset GPIO\n");
-		return PTR_ERR(ctx->gpios.reset);
+		dev_warn(&dsi->dev, "Couldn't get reset GPIO\n");
 	}
 	
 	ctx->gpios.power = devm_gpiod_get(&dsi->dev, "power", GPIOD_OUT_LOW);
 	if (IS_ERR(ctx->gpios.power)) {
-		dev_err(&dsi->dev, "Couldn't get our power GPIO\n");
-		return PTR_ERR(ctx->gpios.power);
+		dev_warn(&dsi->dev, "Couldn't get power GPIO\n");
 	}
 
 	ctx->gpios.updn = devm_gpiod_get_optional(&dsi->dev, "updn", GPIOD_OUT_LOW);
