diff --git a/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c b/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
index 5fba85b52240..4581f26e11ea 100644
--- a/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
+++ b/drivers/gpu/drm/panel/panel-disen-ds070boe30t.c
@@ -204,7 +204,7 @@ static ssize_t read_updn(struct device *dev, struct device_attribute *attr, char
 	}
 
 	/* Check GPIO availability */
-	if (IS_ERR(ctx->gpios.updn))
+	if (IS_ERR(ctx->gpios.updn) || (ctx->gpios.updn == NULL))
 	{
 		dev_err(dev, "<%s> updn GPIO not available\n", __func__);
 		return -EINVAL;
@@ -227,7 +227,7 @@ static ssize_t write_updn(struct device *dev, struct device_attribute *attr, con
 	}
 
 	/* Check GPIO availability */
-	if (IS_ERR(ctx->gpios.updn))
+	if (IS_ERR(ctx->gpios.updn) || (ctx->gpios.updn == NULL))
 	{
 		dev_err(dev, "<%s> updn GPIO not available\n", __func__);
 		return -EINVAL;
@@ -259,7 +259,7 @@ static ssize_t read_shlr(struct device *dev, struct device_attribute *attr, char
 	}
 
 	/* Check GPIO availability */
-	if (IS_ERR(ctx->gpios.shlr))
+	if (IS_ERR(ctx->gpios.shlr) || (ctx->gpios.shlr == NULL))
 	{
 		dev_err(dev, "<%s> shlr GPIO not available\n", __func__);
 		return -EINVAL;
@@ -282,7 +282,7 @@ static ssize_t write_shlr(struct device *dev, struct device_attribute *attr, con
 	}
 
 	/* Check GPIO availability */
-	if (IS_ERR(ctx->gpios.shlr))
+	if (IS_ERR(ctx->gpios.shlr) || (ctx->gpios.shlr == NULL))
 	{
 		dev_err(dev, "<%s> updn GPIO not available\n", __func__);
 		return -EINVAL;
@@ -336,14 +336,14 @@ static int ds070boe30t_panel_dsi_probe(struct mipi_dsi_device *dsi)
 	}
 
 	ctx->gpios.updn = devm_gpiod_get_optional(&dsi->dev, "updn", GPIOD_OUT_LOW);
-	if (IS_ERR(ctx->gpios.updn))
+	if (IS_ERR(ctx->gpios.updn) || (ctx->gpios.updn == NULL))
 		dev_warn(&dsi->dev, "Couldn't get updn GPIO: function won't be available\n");
 	else
 		device_create_file(&dsi->dev, &dev_attr_updn);
 
 	ctx->gpios.shlr = devm_gpiod_get_optional(&dsi->dev, "shlr", GPIOD_OUT_HIGH);
-	if (IS_ERR(ctx->gpios.shlr))
-		dev_warn(&dsi->dev, "Couldn't get shlr GPIO: function won't be availablel\n");
+	if (IS_ERR(ctx->gpios.shlr) || (ctx->gpios.shlr == NULL))
+		dev_warn(&dsi->dev, "Couldn't get shlr GPIO: function won't be available\n");
 	else
 		device_create_file(&dsi->dev, &dev_attr_shlr);
 
