Upstream-Status: Pending

diff --git a/drivers/rtc/rtc-pcf85063.c b/drivers/rtc/rtc-pcf85063.c
index f643e0bd7351..a42aba28d456 100644
--- a/drivers/rtc/rtc-pcf85063.c
+++ b/drivers/rtc/rtc-pcf85063.c
@@ -17,7 +17,6 @@
 #include <linux/of.h>
 #include <linux/pm_wakeirq.h>
 #include <linux/regmap.h>
-#include <linux/spi/spi.h>
 
 /*
  * Information for this driver was pulled from the following datasheets.
@@ -30,16 +29,12 @@
  *
  *  https://www.microcrystal.com/fileadmin/Media/Products/RTC/App.Manual/RV-8263-C7_App-Manual.pdf
  *  RV8263 -- Rev. 1.0 — January 2019
- *
- *  https://www.microcrystal.com/fileadmin/Media/Products/RTC/App.Manual/RV-8063-C7_App-Manual.pdf
- *  RV8063 -- Rev. 1.1 - October 2018
  */
 
 #define PCF85063_REG_CTRL1		0x00 /* status */
 #define PCF85063_REG_CTRL1_CAP_SEL	BIT(0)
 #define PCF85063_REG_CTRL1_STOP		BIT(5)
 #define PCF85063_REG_CTRL1_EXT_TEST	BIT(7)
-#define PCF85063_REG_CTRL1_SWR		0x58
 
 #define PCF85063_REG_CTRL2		0x01
 #define PCF85063_CTRL2_AF		BIT(6)
@@ -327,16 +322,7 @@ static const struct rtc_class_ops pcf85063_rtc_ops = {
 static int pcf85063_nvmem_read(void *priv, unsigned int offset,
 			       void *val, size_t bytes)
 {
-	unsigned int tmp;
-	int ret;
-
-	ret = regmap_read(priv, PCF85063_REG_RAM, &tmp);
-	if (ret < 0)
-		return ret;
-
-	*(u8 *)val = tmp;
-
-	return 0;
+	return regmap_read(priv, PCF85063_REG_RAM, val);
 }
 
 static int pcf85063_nvmem_write(void *priv, unsigned int offset,
@@ -405,19 +391,14 @@ static unsigned long pcf85063_clkout_recalc_rate(struct clk_hw *hw,
 	return clkout_rates[buf];
 }
 
-static int pcf85063_clkout_determine_rate(struct clk_hw *hw,
-					  struct clk_rate_request *req)
+static long pcf85063_clkout_round_rate(struct clk_hw *hw, unsigned long rate,
+				       unsigned long *prate)
 {
 	int i;
 
 	for (i = 0; i < ARRAY_SIZE(clkout_rates); i++)
-		if (clkout_rates[i] <= req->rate) {
-			req->rate = clkout_rates[i];
-
-			return 0;
-		}
-
-	req->rate = clkout_rates[0];
+		if (clkout_rates[i] <= rate)
+			return clkout_rates[i];
 
 	return 0;
 }
@@ -437,9 +418,8 @@ static int pcf85063_clkout_set_rate(struct clk_hw *hw, unsigned long rate,
 	return -EINVAL;
 }
 
-static int pcf85063_clkout_control(struct clk_hw *hw, bool enable)
+static int pcf85063_clkout_control_pcf(struct pcf85063 *pcf85063, bool enable)
 {
-	struct pcf85063 *pcf85063 = clkout_hw_to_pcf85063(hw);
 	unsigned int buf;
 	int ret;
 
@@ -464,6 +444,11 @@ static int pcf85063_clkout_control(struct clk_hw *hw, bool enable)
 					PCF85063_REG_CLKO_F_MASK, buf);
 }
 
+static int pcf85063_clkout_control(struct clk_hw *hw, bool enable)
+{
+	return(pcf85063_clkout_control_pcf(clkout_hw_to_pcf85063(hw), enable));
+}
+
 static int pcf85063_clkout_prepare(struct clk_hw *hw)
 {
 	return pcf85063_clkout_control(hw, 1);
@@ -474,9 +459,8 @@ static void pcf85063_clkout_unprepare(struct clk_hw *hw)
 	pcf85063_clkout_control(hw, 0);
 }
 
-static int pcf85063_clkout_is_prepared(struct clk_hw *hw)
+static int pcf85063_clkout_is_prepared_pcf(struct pcf85063 *pcf85063)
 {
-	struct pcf85063 *pcf85063 = clkout_hw_to_pcf85063(hw);
 	unsigned int buf;
 	int ret = regmap_read(pcf85063->regmap, PCF85063_REG_CTRL2, &buf);
 
@@ -486,12 +470,17 @@ static int pcf85063_clkout_is_prepared(struct clk_hw *hw)
 	return (buf & PCF85063_REG_CLKO_F_MASK) != PCF85063_REG_CLKO_F_OFF;
 }
 
+static int pcf85063_clkout_is_prepared(struct clk_hw *hw)
+{
+	return(pcf85063_clkout_is_prepared_pcf(clkout_hw_to_pcf85063(hw)));
+}
+
 static const struct clk_ops pcf85063_clkout_ops = {
 	.prepare = pcf85063_clkout_prepare,
 	.unprepare = pcf85063_clkout_unprepare,
 	.is_prepared = pcf85063_clkout_is_prepared,
 	.recalc_rate = pcf85063_clkout_recalc_rate,
-	.determine_rate = pcf85063_clkout_determine_rate,
+	.round_rate = pcf85063_clkout_round_rate,
 	.set_rate = pcf85063_clkout_set_rate,
 };
 
@@ -510,6 +499,10 @@ static struct clk *pcf85063_clkout_register_clk(struct pcf85063 *pcf85063)
 		 * registered automatically when being referenced.
 		 */
 		of_node_put(fixed_clock);
+
+		/* Determine initial status of clkout */
+		pcf85063_clkout_control_pcf(pcf85063, !of_property_read_bool(fixed_clock, "disabled-at-start"));
+
 		return NULL;
 	}
 
@@ -533,12 +526,108 @@ static struct clk *pcf85063_clkout_register_clk(struct pcf85063 *pcf85063)
 }
 #endif
 
-static int pcf85063_probe(struct device *dev, struct regmap *regmap, int irq,
-			  const struct pcf85063_config *config)
+static const struct pcf85063_config config_pcf85063 = {
+	.regmap = {
+		.reg_bits = 8,
+		.val_bits = 8,
+		.max_register = 0x0a,
+	},
+};
+
+static const struct pcf85063_config config_pcf85063tp = {
+	.regmap = {
+		.reg_bits = 8,
+		.val_bits = 8,
+		.max_register = 0x0a,
+	},
+};
+
+static const struct pcf85063_config config_pcf85063a = {
+	.regmap = {
+		.reg_bits = 8,
+		.val_bits = 8,
+		.max_register = 0x11,
+	},
+	.has_alarms = 1,
+};
+
+static const struct pcf85063_config config_rv8263 = {
+	.regmap = {
+		.reg_bits = 8,
+		.val_bits = 8,
+		.max_register = 0x11,
+	},
+	.has_alarms = 1,
+	.force_cap_7000 = 1,
+};
+
+static ssize_t read_fixed_clock_disable(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	struct pcf85063 *pcf85063 = dev_get_drvdata(dev);
+
+	if(!pcf85063)
+	{
+		dev_err(dev, "<%s> can't get driver data\n", __func__);
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_COMMON_CLK
+	if(pcf85063->clkout_hw.init)
+	{
+		dev_err(dev, "<%s> not a fixed clock: cannot read fixed clock status\n", __func__);
+		return -EINVAL;
+	}
+#endif
+
+	/* Set data */
+	return scnprintf(buf, PAGE_SIZE, "%d\n", (!pcf85063_clkout_is_prepared_pcf(pcf85063) ? 1 : 0));
+}
+
+static ssize_t write_fixed_clock_disable(struct device *dev, struct device_attribute *attr, const char *buf, size_t count)
+{
+	struct pcf85063 *pcf85063 = dev_get_drvdata(dev);
+
+	if(!pcf85063)
+	{
+		dev_err(dev, "<%s> can't get driver data\n", __func__);
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_COMMON_CLK
+	if(pcf85063->clkout_hw.init)
+	{
+		dev_err(dev, "<%s> not a fixed clock: cannot set fixed clock status\n", __func__);
+		return -EINVAL;
+	}
+#endif
+
+	/* Enforce status */
+	pcf85063_clkout_control_pcf(pcf85063, !(count > 0 && buf[0] == '1'));
+
+	/* Dump info */
+	dev_info(dev, "<%s> fixed_clock_disable set to %d\n", __func__, pcf85063_clkout_is_prepared_pcf(pcf85063));
+
+	return count;
+}
+
+static DEVICE_ATTR(fixed_clock_disable, 0644, read_fixed_clock_disable, write_fixed_clock_disable);
+
+static int pcf85063_sysfs_register(struct device *dev)
+{
+	return device_create_file(dev, &dev_attr_fixed_clock_disable);
+}
+
+static void pcf85063_sysfs_unregister(struct device *dev)
+{
+	device_remove_file(dev, &dev_attr_fixed_clock_disable);
+}
+
+static int pcf85063_probe(struct i2c_client *client)
 {
 	struct pcf85063 *pcf85063;
 	unsigned int tmp;
 	int err;
+	const struct pcf85063_config *config;
 	struct nvmem_config nvmem_cfg = {
 		.name = "pcf85063_nvram",
 		.reg_read = pcf85063_nvmem_read,
@@ -547,43 +636,37 @@ static int pcf85063_probe(struct device *dev, struct regmap *regmap, int irq,
 		.size = 1,
 	};
 
-	dev_dbg(dev, "%s\n", __func__);
+	dev_dbg(&client->dev, "%s\n", __func__);
 
-	pcf85063 = devm_kzalloc(dev, sizeof(struct pcf85063),
+	pcf85063 = devm_kzalloc(&client->dev, sizeof(struct pcf85063),
 				GFP_KERNEL);
 	if (!pcf85063)
 		return -ENOMEM;
 
-	pcf85063->regmap = regmap;
+	config = i2c_get_match_data(client);
+	if (!config)
+		return -ENODEV;
 
-	dev_set_drvdata(dev, pcf85063);
+	pcf85063->regmap = devm_regmap_init_i2c(client, &config->regmap);
+	if (IS_ERR(pcf85063->regmap))
+		return PTR_ERR(pcf85063->regmap);
 
-	err = regmap_read(pcf85063->regmap, PCF85063_REG_SC, &tmp);
-	if (err)
-		return dev_err_probe(dev, err, "RTC chip is not present\n");
+	i2c_set_clientdata(client, pcf85063);
 
-	pcf85063->rtc = devm_rtc_allocate_device(dev);
+	err = regmap_read(pcf85063->regmap, PCF85063_REG_CTRL1, &tmp);
+	if (err) {
+		dev_err(&client->dev, "RTC chip is not present\n");
+		return err;
+	}
+
+	pcf85063->rtc = devm_rtc_allocate_device(&client->dev);
 	if (IS_ERR(pcf85063->rtc))
 		return PTR_ERR(pcf85063->rtc);
 
-	/*
-	 * If a Power loss is detected, SW reset the device.
-	 * From PCF85063A datasheet:
-	 * There is a low probability that some devices will have corruption
-	 * of the registers after the automatic power-on reset...
-	 */
-	if (tmp & PCF85063_REG_SC_OS) {
-		dev_warn(dev, "POR issue detected, sending a SW reset\n");
-		err = regmap_write(pcf85063->regmap, PCF85063_REG_CTRL1,
-				   PCF85063_REG_CTRL1_SWR);
-		if (err < 0)
-			dev_warn(dev, "SW reset failed, trying to continue\n");
-	}
-
-	err = pcf85063_load_capacitance(pcf85063, dev->of_node,
+	err = pcf85063_load_capacitance(pcf85063, client->dev.of_node,
 					config->force_cap_7000 ? 7000 : 0);
 	if (err < 0)
-		dev_warn(dev, "failed to set xtal load capacitance: %d",
+		dev_warn(&client->dev, "failed to set xtal load capacitance: %d",
 			 err);
 
 	pcf85063->rtc->ops = &pcf85063_rtc_ops;
@@ -593,13 +676,13 @@ static int pcf85063_probe(struct device *dev, struct regmap *regmap, int irq,
 	clear_bit(RTC_FEATURE_UPDATE_INTERRUPT, pcf85063->rtc->features);
 	clear_bit(RTC_FEATURE_ALARM, pcf85063->rtc->features);
 
-	if (config->has_alarms && irq > 0) {
+	if (config->has_alarms && client->irq > 0) {
 		unsigned long irqflags = IRQF_TRIGGER_LOW;
 
-		if (dev_fwnode(dev))
+		if (dev_fwnode(&client->dev))
 			irqflags = 0;
 
-		err = devm_request_threaded_irq(dev, irq,
+		err = devm_request_threaded_irq(&client->dev, client->irq,
 						NULL, pcf85063_rtc_handle_irq,
 						irqflags | IRQF_ONESHOT,
 						"pcf85063", pcf85063);
@@ -608,8 +691,8 @@ static int pcf85063_probe(struct device *dev, struct regmap *regmap, int irq,
 				 "unable to request IRQ, alarms disabled\n");
 		} else {
 			set_bit(RTC_FEATURE_ALARM, pcf85063->rtc->features);
-			device_init_wakeup(dev, true);
-			err = dev_pm_set_wake_irq(dev, irq);
+			device_init_wakeup(&client->dev, true);
+			err = dev_pm_set_wake_irq(&client->dev, client->irq);
 			if (err)
 				dev_err(&pcf85063->rtc->dev,
 					"failed to enable irq wake\n");
@@ -624,45 +707,23 @@ static int pcf85063_probe(struct device *dev, struct regmap *regmap, int irq,
 	pcf85063_clkout_register_clk(pcf85063);
 #endif
 
-	return devm_rtc_register_device(pcf85063->rtc);
-}
-
-#if IS_ENABLED(CONFIG_I2C)
-
-static const struct pcf85063_config config_pcf85063 = {
-	.regmap = {
-		.reg_bits = 8,
-		.val_bits = 8,
-		.max_register = 0x0a,
-	},
-};
+	err = devm_rtc_register_device(pcf85063->rtc);
+	if (err)
+		return err;
 
-static const struct pcf85063_config config_pcf85063tp = {
-	.regmap = {
-		.reg_bits = 8,
-		.val_bits = 8,
-		.max_register = 0x0a,
-	},
-};
+	err = pcf85063_sysfs_register(&client->dev);
+	if (err) {
+		dev_err(&client->dev, "Unable to create sysfs entries for rtc pcf85063\n");
+		return(err);
+	}
 
-static const struct pcf85063_config config_pcf85063a = {
-	.regmap = {
-		.reg_bits = 8,
-		.val_bits = 8,
-		.max_register = 0x11,
-	},
-	.has_alarms = 1,
-};
+	return(0);
+}
 
-static const struct pcf85063_config config_rv8263 = {
-	.regmap = {
-		.reg_bits = 8,
-		.val_bits = 8,
-		.max_register = 0x11,
-	},
-	.has_alarms = 1,
-	.force_cap_7000 = 1,
-};
+static void pcf85063_remove(struct i2c_client *client)
+{
+	pcf85063_sysfs_unregister(&client->dev);
+}
 
 static const struct i2c_device_id pcf85063_ids[] = {
 	{ "pca85073a", .driver_data = (kernel_ulong_t)&config_pcf85063a },
@@ -686,146 +747,17 @@ static const struct of_device_id pcf85063_of_match[] = {
 MODULE_DEVICE_TABLE(of, pcf85063_of_match);
 #endif
 
-static int pcf85063_i2c_probe(struct i2c_client *client)
-{
-	const struct pcf85063_config *config;
-	struct regmap *regmap;
-
-	config = i2c_get_match_data(client);
-	if (!config)
-		return -ENODEV;
-
-	regmap = devm_regmap_init_i2c(client, &config->regmap);
-	if (IS_ERR(regmap))
-		return PTR_ERR(regmap);
-
-	return pcf85063_probe(&client->dev, regmap, client->irq, config);
-}
-
 static struct i2c_driver pcf85063_driver = {
 	.driver		= {
 		.name	= "rtc-pcf85063",
 		.of_match_table = of_match_ptr(pcf85063_of_match),
 	},
-	.probe		= pcf85063_i2c_probe,
+	.probe		= pcf85063_probe,
+	.remove		= pcf85063_remove,
 	.id_table	= pcf85063_ids,
 };
 
-static int pcf85063_register_driver(void)
-{
-	return i2c_add_driver(&pcf85063_driver);
-}
-
-static void pcf85063_unregister_driver(void)
-{
-	i2c_del_driver(&pcf85063_driver);
-}
-
-#else
-
-static int pcf85063_register_driver(void)
-{
-	return 0;
-}
-
-static void pcf85063_unregister_driver(void)
-{
-}
-
-#endif /* IS_ENABLED(CONFIG_I2C) */
-
-#if IS_ENABLED(CONFIG_SPI_MASTER)
-
-static const struct pcf85063_config config_rv8063 = {
-	.regmap = {
-		.reg_bits = 8,
-		.val_bits = 8,
-		.max_register = 0x11,
-		.read_flag_mask = BIT(7) | BIT(5),
-		.write_flag_mask = BIT(5),
-	},
-	.has_alarms = 1,
-	.force_cap_7000 = 1,
-};
-
-static const struct spi_device_id rv8063_id[] = {
-	{ "rv8063" },
-	{}
-};
-MODULE_DEVICE_TABLE(spi, rv8063_id);
-
-static const struct of_device_id rv8063_of_match[] = {
-	{ .compatible = "microcrystal,rv8063" },
-	{}
-};
-MODULE_DEVICE_TABLE(of, rv8063_of_match);
-
-static int rv8063_probe(struct spi_device *spi)
-{
-	const struct pcf85063_config *config = &config_rv8063;
-	struct regmap *regmap;
-
-	regmap = devm_regmap_init_spi(spi, &config->regmap);
-	if (IS_ERR(regmap))
-		return PTR_ERR(regmap);
-
-	return pcf85063_probe(&spi->dev, regmap, spi->irq, config);
-}
-
-static struct spi_driver rv8063_driver = {
-	.driver         = {
-		.name   = "rv8063",
-		.of_match_table = rv8063_of_match,
-	},
-	.probe          = rv8063_probe,
-	.id_table	= rv8063_id,
-};
-
-static int __init rv8063_register_driver(void)
-{
-	return spi_register_driver(&rv8063_driver);
-}
-
-static void __exit rv8063_unregister_driver(void)
-{
-	spi_unregister_driver(&rv8063_driver);
-}
-
-#else
-
-static int __init rv8063_register_driver(void)
-{
-	return 0;
-}
-
-static void __exit rv8063_unregister_driver(void)
-{
-}
-
-#endif /* IS_ENABLED(CONFIG_SPI_MASTER) */
-
-static int __init pcf85063_init(void)
-{
-	int ret;
-
-	ret = pcf85063_register_driver();
-	if (ret)
-		return ret;
-
-	ret = rv8063_register_driver();
-	if (ret)
-		pcf85063_unregister_driver();
-
-	return ret;
-}
-module_init(pcf85063_init);
-
-static void __exit pcf85063_exit(void)
-{
-	rv8063_unregister_driver();
-	pcf85063_unregister_driver();
-}
-module_exit(pcf85063_exit);
+module_i2c_driver(pcf85063_driver);
 
 MODULE_AUTHOR("Søren Andersen <san@rosetechnology.dk>");
 MODULE_DESCRIPTION("PCF85063 RTC driver");
