Upstream-Status: Pending
Signed-off-by: Morix Dev <morix.dev@gmail.com>


diff --git a/drivers/clk/imx/clk-pll14xx.c b/drivers/clk/imx/clk-pll14xx.c
index 6c17786ecb9f..88701532380a 100644
--- a/drivers/clk/imx/clk-pll14xx.c
+++ b/drivers/clk/imx/clk-pll14xx.c
@@ -56,7 +56,10 @@ static const struct imx_pll14xx_rate_table imx_pll1416x_tbl[] = {
 	PLL_1416X_RATE(700000000U,  350, 3, 2),
 	PLL_1416X_RATE(640000000U,  320, 3, 2),
 	PLL_1416X_RATE(600000000U,  300, 3, 2),
+    PLL_1416X_RATE(400000000U,  200, 3, 2),
 	PLL_1416X_RATE(320000000U,  160, 3, 2),
+    PLL_1416X_RATE(200000000U,  200, 3, 3),
+    PLL_1416X_RATE(100000000U,  200, 3, 4),
 };
 
 static const struct imx_pll14xx_rate_table imx_pll1443x_tbl[] = {
