Upstream-Status: Pending
Signed-off-by: Morix Dev <morix.dev@gmail.com>



diff --git a/disk/part.c b/disk/part.c
index 36b88205eca..409c56590eb 100644
--- a/disk/part.c
+++ b/disk/part.c
@@ -633,7 +633,8 @@ int blk_get_device_part_str(const char *ifname, const char *dev_part_str,
 	if (part != PART_AUTO) {
 		ret = part_get_info(*desc, part, info);
 		if (ret) {
-			printf("** Invalid partition %d **\n", part);
+            if(!env_get("quiet_part"))
+			    printf("** Invalid partition %d **\n", part);
 			goto cleanup;
 		}
 	} else {
@@ -800,9 +801,10 @@ int part_get_info_by_dev_and_name_or_num(const char *dev_iface,
 	 */
 	ret = blk_get_device_part_str(dev_iface, dev_part_str, desc, part_info,
 				      allow_whole_dev);
-	if (ret < 0)
-		printf("Couldn't find partition %s %s\n",
-		       dev_iface, dev_part_str);
+	if (ret < 0) {
+        if(!env_get("quiet_part"))
+		    printf("Couldn't find partition %s %s\n", dev_iface, dev_part_str);
+    }
 	return ret;
 }
 
