--- a/media/gpu/args.gni
+++ b/media/gpu/args.gni
@@ -6,6 +6,11 @@
 import("//build/config/ozone.gni")
 
 declare_args() {
+  # Checks if i.MX GPU is being used
+  use_imxgpu = false
+}
+
+declare_args() {
   # Indicates if X11 VA-API-based hardware acceleration is to be used.
   # See also the comment near the |use_vaapi| arg.
   use_vaapi_x11 = is_linux && ozone_platform_x11 &&
--- a/media/gpu/sandbox/BUILD.gn
+++ b/media/gpu/sandbox/BUILD.gn
@@ -31,7 +31,7 @@
     deps += [ "//media/gpu/v4l2" ]
   }
   if (current_cpu != "s390x" && current_cpu != "ppc64" && is_linux &&
-      !is_castos) {
+      !is_castos && !use_imxgpu) {
     # For DRI_DRIVER_DIR.
     configs += [ "//build/config/linux/dri" ]
   }
--- a/content/gpu/BUILD.gn
+++ b/content/gpu/BUILD.gn
@@ -7,6 +7,7 @@
 import("//build/config/ui.gni")
 import("//gpu/vulkan/features.gni")
 import("//media/media_options.gni")
+import("//media/gpu/args.gni")
 
 # See //content/BUILD.gn for how this works.
 group("gpu") {
@@ -140,7 +141,7 @@
 
   # Use DRI on desktop Linux builds.
   if (current_cpu != "s390x" && current_cpu != "ppc64" && is_linux &&
-      !is_castos) {
+      !is_castos && !use_imxgpu) {
     configs += [ "//build/config/linux/dri" ]
   }
 }
